File: backend/api-gateway/internal/clients/auth_client.go
// Package clients envuelve los clientes gRPC hacia los microservicios.
// Aqui esta el cliente del auth-service.
package clients

import (
	"context"

	"github.com/grupo4/quetxaltv-gateway/internal/pb"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
)

type AuthClient struct {
	conn *grpc.ClientConn
	cli  pb.AuthServiceClient
}

// NewAuthClient abre la conexion gRPC con el auth-service.
// Usa credenciales 'insecure' porque es trafico interno entre servicios
// dentro de la red de Docker (sin TLS).
func NewAuthClient(addr string) (*AuthClient, error) {
	conn, err := grpc.NewClient(
		addr,
		grpc.WithTransportCredentials(insecure.NewCredentials()),
		grpc.WithUnaryInterceptor(MetadataForwardingInterceptor),
	)
	if err != nil {
		return nil, err
	}
	return &AuthClient{conn: conn, cli: pb.NewAuthServiceClient(conn)}, nil
}

func (c *AuthClient) Close() error { return c.conn.Close() }

func (c *AuthClient) Registrar(ctx context.Context, email, password, nombrePerfil string) (*pb.RegistrarResponse, error) {
	return c.cli.Registrar(ctx, &pb.RegistrarRequest{
		Email: email, Password: password, NombrePerfil: nombrePerfil,
	})
}

func (c *AuthClient) Login(ctx context.Context, email, password string) (*pb.LoginResponse, error) {
	return c.cli.Login(ctx, &pb.LoginRequest{Email: email, Password: password})
}

func (c *AuthClient) ValidarToken(ctx context.Context, token string) (*pb.ValidarTokenResponse, error) {
	return c.cli.ValidarToken(ctx, &pb.ValidarTokenRequest{Token: token})
}

func (c *AuthClient) CrearPerfil(ctx context.Context, usuarioID, nombre, idioma string, esInfantil bool, pin string) (*pb.PerfilResponse, error) {
	return c.cli.CrearPerfil(ctx, &pb.CrearPerfilRequest{
		UsuarioId: usuarioID, Nombre: nombre, Idioma: idioma, EsInfantil: esInfantil, Pin: pin,
	})
}

func (c *AuthClient) EditarPerfil(ctx context.Context, usuarioID, perfilID, nombre, idioma string, esInfantil bool, pin string) (*pb.PerfilResponse, error) {
	return c.cli.EditarPerfil(ctx, &pb.EditarPerfilRequest{
		UsuarioId: usuarioID, PerfilId: perfilID, Nombre: nombre, Idioma: idioma, EsInfantil: esInfantil, Pin: pin,
	})
}

func (c *AuthClient) ListarPerfiles(ctx context.Context, usuarioID string) (*pb.ListarPerfilesResponse, error) {
	return c.cli.ListarPerfiles(ctx, &pb.ListarPerfilesRequest{UsuarioId: usuarioID})
}

func (c *AuthClient) ActualizarPerfil(ctx context.Context, id, usuarioID, nombre string) (*pb.PerfilResponse, error) {
	return c.cli.ActualizarPerfil(ctx, &pb.ActualizarPerfilRequest{
		Id: id, UsuarioId: usuarioID, Nombre: nombre,
	})
}

func (c *AuthClient) EliminarPerfil(ctx context.Context, id, usuarioID string) (*pb.EliminarPerfilResponse, error) {
	return c.cli.EliminarPerfil(ctx, &pb.EliminarPerfilRequest{
		Id: id, UsuarioId: usuarioID,
	})
}

File: frontend/lib/grpc/auth-client.ts
import * as grpc from '@grpc/grpc-js'
import * as protoLoader from '@grpc/proto-loader'
import path from 'path'

const PROTO_PATH = path.join(process.cwd(), 'proto', 'auth.proto')

const packageDefinition = protoLoader.loadSync(PROTO_PATH, {
  keepCase: false,
  longs: Number,
  enums: String,
  defaults: true,
  oneofs: true,
})

const proto = grpc.loadPackageDefinition(packageDefinition) as any

const AUTH_SERVICE_URL = process.env.GRPC_AUTH_URL || 'localhost:50051'

let client: any = null

function getClient() {
  if (!client) {
    client = new proto.auth.v1.AuthService(
      AUTH_SERVICE_URL,
      grpc.credentials.createInsecure()
    )
  }
  return client
}

function promisify<T>(fn: Function, request: any): Promise<T> {
  return new Promise((resolve, reject) => {
    fn.call(getClient(), request, (err: grpc.ServiceError | null, response: T) => {
      if (err) reject(err)
      else resolve(response)
    })
  })
}

// --- Tipos ---

export interface RegistrarRequest {
  email: string
  password: string
  nombrePerfil: string
}

export interface RegistrarResponse {
  usuarioId: string
  perfilId: string
}

export interface LoginRequest {
  email: string
  password: string
}

export interface LoginResponse {
  accessToken: string
  usuarioId: string
  expiraEn: number
}

export interface ValidarTokenRequest {
  token: string
}

export interface ValidarTokenResponse {
  valido: boolean
  usuarioId: string
  rol: string
}

export interface CrearPerfilRequest {
  usuarioId: string
  nombre: string
  esInfantil: boolean
  idioma: string
}

export interface PerfilResponse {
  id: string
  nombre: string
  esInfantil: boolean
  idioma: string
}

export interface ListarPerfilesResponse {
  perfiles: PerfilResponse[]
}

// --- Funciones del cliente ---

export async function registrar(req: RegistrarRequest): Promise<RegistrarResponse> {
  return promisify<RegistrarResponse>(getClient().registrar, {
    email: req.email,
    password: req.password,
    nombrePerfil: req.nombrePerfil,
  })
}

export async function login(req: LoginRequest): Promise<LoginResponse> {
  return promisify<LoginResponse>(getClient().login, {
    email: req.email,
    password: req.password,
  })
}

export async function validarToken(token: string): Promise<ValidarTokenResponse> {
  return promisify<ValidarTokenResponse>(getClient().validarToken, { token })
}

export async function crearPerfil(req: CrearPerfilRequest): Promise<PerfilResponse> {
  return promisify<PerfilResponse>(getClient().crearPerfil, {
    usuarioId: req.usuarioId,
    nombre: req.nombre,
    esInfantil: req.esInfantil,
    idioma: req.idioma,
  })
}

export async function listarPerfiles(usuarioId: string): Promise<ListarPerfilesResponse> {
  return promisify<ListarPerfilesResponse>(getClient().listarPerfiles, {
    usuarioId: usuarioId,
  })
}
